### PR TITLE
speed up hole detection by 400x

### PR DESCRIPTION
**Pull request by staltz**
I had a hunch that `buf.every((x) => x === 0)` is probably slow, so I did some research, and ended up finding this lovely module by mafintosh: https://github.com/mafintosh/is-zero-buffer

I ran these simple benchmarks:

```js
var isZero = (buf) => buf.every(x => x === 0)

var buf = Buffer.alloc(65536)
var then = Date.now()

for (var i = 0; i < 100000; i++) {
  isZero(buf)
}

console.log(Date.now() - then)
```

```js
var isZero = require('is-zero-buffer')

var buf = Buffer.alloc(65536)
var then = Date.now()

for (var i = 0; i < 100000; i++) {
  isZero(buf)
}

console.log(Date.now() - then)
```

And the numbers are ridiculous:

- ~`99491` milliseconds
- ~`225` milliseconds

Easy win!

### CODE REVIEW

**Review by arj03 — APPROVED** [review]
Interesting. Wouldn't have thought there was such a big difference. Records are at most 8k, but still, this is quite an improvement for deleted data.

**Comment by arj03** [comment]
Can you try running this microbenchmark with random data?

I'm seeing this in jitdb:

Before:

```
  query one huge index (first run)

    ✔ | Query 1 big index (1st run) | 1199.9ms ± 56.76ms | -91.96 kB ± 162.02 kB | 5 | 

  query one huge index (second run)

    ✔ | Query 1 big index (2nd run) | 358.25ms ± 2.43ms | 93.54 kB ± 248.12 kB | 5 | 
```

After:

```
  query one huge index (first run)

    ✔ | Query 1 big index (1st run) | 1347.25ms ± 41.9ms | -41.5 kB ± 593.94 kB | 5 | 

  query one huge index (second run)

    ✔ | Query 1 big index (2nd run) | 409.53ms ± 6.4ms | -11.49 kB ± 455.79 kB | 5 | 
```

**Comment by staltz** [comment]
> Can you try running this microbenchmark with random data?

Sure, I'll try that.

> Records are at most 8k

I tried this microbenchmark with buffer size 8192 and the outcome was `63657 ms` versus `133 ms`, so about the same order of magnitude for the speed up.

**Comment by staltz** [comment]
Wow, I think you're onto something.

```js
var crypto = require('crypto');
// var isZero = require('./')
var isZero = (buf) => buf.every((x) => x === 0);

var MAX_SIZE = 8192
var TOTAL = 125000
var buffers = []
for (var i = 0; i < TOTAL; i++) {
  const size = Math.ceil(Math.random() * MAX_SIZE)
  buffers.push(crypto.randomBytes(size))
}

var then = Date.now()
for (var i = 0; i < TOTAL; i++) {
  isZero(buffers[i])
}
console.log(Date.now() - then)
```

## `every(x => x === 0)`

Ten different runs of the script:

35
35
55
43
91
55
57
35
35
35

Average 47ms

## isZeroBuf

27
18
19
39
39
38
18
39
29
38
18

Average 30ms

**Comment by staltz** [comment]
I was trying to think what's the explanation for this, and it seems to be that when the record is "real" (not deleted), then `every(x => x === 0)` will bail out quickly, on the 1st or 2nd byte, most of the times. However, if the record is deleted, then `x => x === 0` will have to traverse **all** the bytes, while `buffer.compare` (what is-zero-buffer uses under the hood) has some advantage in that case.

In other words, I think the conclusion is that we should use `is-zero-buffer` only when most records are deleted, while we should use `every` when most records are real. Reasonable to assume we have the latter case?

**Comment by staltz** [comment]
Hmmm, I wonder if there is a compromise. Some kind of `is-zero-buffer` fork that by default uses a for loop over the bytes, but when it detects at least N bytes of the first bytes are zero, then "probably" the whole buffer is zeroes, and it would use `compare` instead.

What do you think? Right now, this kind of optimization is not too important, but we can make an issue to remind ourselves, and when we are in production where we have real world logs with a bunch of deleted records, then we can macrobenchmark this better.

**Comment by arj03** [comment]
Yep, a fork would be best I think. The api would be something like peekSize where you could define the number of bytes to peek for zeroes, before falling back to the check is-zero-buffer does. The module is extremely simple so should be easy :)

**Comment by staltz** [comment]
Good. I'll close this PR and make the issue to remind myself about this :)
